File: term/term/commands.py
"""Catalogo de comandos, atajos y prompt de sistema."""

from __future__ import annotations

from .i18n import LANGUAGES

__all__ = [
    "COMMANDS_HELP",
    "COMMAND_NAMES",
    "SHORTCUTS_HELP",
    "build_system_context",
    "complete_command",
]

# Los comandos van agrupados porque la ayuda los enseña asi: cincuenta
# comandos en una sola lista son un muro que nadie lee. La clave de cada grupo
# es la de su titulo traducido.
COMMAND_GROUPS: dict[str, dict[str, str]] = {
    "grp_chat": {
        "/new [nombre] [modelo]": "Nueva pestaña, opcionalmente con su IA",
        "/close":                 "Cerrar la pestaña activa",
        "/clear":                 "Limpiar el chat y empezar una sesión nueva",
        "/sessions":              "Listar las conversaciones guardadas",
        "/resume <n>":            "Retomar una conversación guardada",
        "/search <texto>":        "Buscar dentro de esta conversación",
        "/name <texto>":          "Renombrar la pestaña activa",
        "/history":               "Cuántos mensajes lleva esta pestaña",
        "/export":                "Guardar la conversación en un archivo",
        "/copy":                  "Copiar la última respuesta",
        "/code [n]":              "Copiar el bloque de código n de la última respuesta",
    },
    "grp_ai": {
        "/model <proveedor/modelo>": "Cambiar la IA de esta pestaña",
        "/providers":             "Ver cómo está conectada cada IA",
        "/key <proveedor> <clave>": "Guardar la API key de un proveedor",
        "/key-del <proveedor>":   "Borrar una API key guardada",
        "/effort <nivel>":        "Nivel de esfuerzo (low, medium, high, max)",
        "/architect [modelo]":    "Que otro modelo planifique antes de ejecutar",
        "/skeleton":              "Pasar el esqueleto del código en vez de la lista de archivos",
        "/permissions [modo]":    "Modo de permisos (default, acceptEdits, plan, bypassPermissions)",
        "/allow [perfil]":        "Qué herramientas puede usar la IA (todo, lectura, segura, nada)",
        "/mcp":                   "Servidores MCP y sus herramientas",
        "/mcp-add <nombre> <cmd>": "Añadir un servidor MCP",
        "/mcp-del <nombre>":      "Quitar un servidor MCP",
        "/compact":               "Resumir la conversación para liberar contexto",
    },
    "grp_git": {
        "/status":                "Qué has cambiado en el repositorio",
        "/diff":                  "Ver los cambios sin confirmar",
        "/commit [mensaje]":      "Guardar los cambios; sin mensaje lo escribe la IA",
        "/undo":                  "Deshacer el último commit sin perder el trabajo",
        "/gitlog":                "Últimos commits",
    },
    "grp_files": {
        "/mkdir <ruta>":          "Crear una carpeta, con sus padres si hacen falta",
        "/touch <ruta>":          "Crear un archivo vacío",
        "/find <patrón>":         "Buscar archivos aquí y mostrar sus rutas",
        "/findall <patrón>":      "Buscar en todo el disco con Spotlight",
        "/grep <texto>":          "Buscar un texto dentro de los archivos",
        "/files":                 "Mostrar u ocultar el panel de archivos",
        "/add <ruta>":            "Meter un archivo en el contexto de la conversación",
        "/drop <ruta>":           "Sacar un archivo del contexto",
        "/context":               "Ver qué archivos están en el contexto",
        "/map":                   "Ver el mapa del proyecto",
        "/outline [ruta]":        "Ver clases y funciones con su firma",
        "/attach <ruta>":         "Adjuntar un archivo solo al siguiente mensaje",
        "/detach":                "Descartar los archivos adjuntos pendientes",
        "/workdir <ruta>":        "Cambiar el directorio de trabajo",
    },
    "grp_system": {
        "/run <cmd>":             "Ejecutar un comando y esperar su salida",
        "/bg <cmd>":              "Lanzar un comando en segundo plano (servidor, tests…)",
        "/jobs":                  "Ver los procesos en segundo plano",
        "/logs [n] [texto]":      "Ver la salida de un proceso, filtrando si quieres",
        "/stop [n]":              "Parar un proceso; sin número, todos",
        "/open <app>":            "Abrir una aplicación",
        "/close-app <app>":       "Cerrar una aplicación",
        "/web <consulta>":        "Buscar en Google desde el navegador",
        "/yt <consulta>":         "Buscar en YouTube",
        "/maps <lugar>":          "Buscar un lugar en Google Maps",
        "/browse [url]":          "Abrir una URL en el navegador",
        "/browser <nombre>":      "Elegir el navegador por defecto",
        "/play":                  "Reanudar la música",
        "/pause":                 "Pausar la música",
        "/next":                  "Siguiente canción",
        "/prev":                  "Canción anterior",
        "/track":                 "Qué canción suena ahora",
        "/volume [0-100]":        "Ver o ajustar el volumen",
        "/sysinfo":               "Batería, disco, red, volumen y música",
    },
    "grp_look": {
        "/theme <nombre>":        "Cambiar el tema de color",
        "/lang [código]":         f"Cambiar el idioma ({', '.join(LANGUAGES)})",
        "/settings":              "Panel de ajustes",
        "/apps":                  "Aplicaciones detectadas",
        "/tools":                 "Herramientas detectadas",
        "/help":                  "Esta ayuda",
    },
    "grp_meta": {
        "/tab":                   "Estado de la pestaña activa",
        "/reset":                 "Poner a cero el contador de contexto",
        "/save":                  "Guardar la configuración en disco",
        "/version":               "Versión de Term",
        "/about":                 "Acerca de Term",
        "/quit":                  "Salir",
    },
}

# Vista plana de todos los comandos, que es lo que consultan el autocompletado
# y las sugerencias mientras se escribe.
COMMANDS_HELP: dict[str, str] = {
    cmd: desc for grupo in COMMAND_GROUPS.values() for cmd, desc in grupo.items()
}

# Nombre pelado de cada comando, para autocompletar.
COMMAND_NAMES: list[str] = [c.split()[0] for c in COMMANDS_HELP]

SHORTCUTS_HELP: dict[str, str] = {
    "enter":        "Enviar el mensaje",
    "alt+enter":    "Salto de línea dentro del mensaje",
    "up / down":    "Recorrer los mensajes que ya has enviado",
    "tab":          "Autocompletar el comando que estás escribiendo",
    "ctrl+t":       "Nueva pestaña",
    "ctrl+w":       "Cerrar la pestaña o el panel activo",
    "ctrl+1..9":    "Saltar a la pestaña n",
    "ctrl+l":       "Limpiar el chat",
    "ctrl+e":       "Cambiar el nivel de esfuerzo",
    "ctrl+b":       "Mostrar u ocultar el panel de archivos",
    "ctrl+y":       "Copiar la última respuesta",
    "escape":       "Cancelar la generación o cerrar el panel",
    "ctrl+c":       "Salir",
}


def complete_command(prefix: str) -> tuple[str, list[str]]:
    """Completar un comando a medio escribir.

    Devuelve `(completado, candidatos)`. `completado` es el prefijo comun mas
    largo, para que pulsar tab avance aunque haya varias opciones.
    """
    prefix = prefix.strip()
    if not prefix.startswith("/"):
        return prefix, []
    matches = [c for c in COMMAND_NAMES if c.startswith(prefix)]
    if not matches:
        return prefix, []
    if len(matches) == 1:
        return matches[0] + " ", matches

    common = matches[0]
    for candidate in matches[1:]:
        while not candidate.startswith(common):
            common = common[:-1]
    return common, matches


_SYSTEM_BASE = (
    "Eres Term, un asistente que vive en una TUI de terminal.\n"
    "Tienes acceso al shell del usuario a través de tus herramientas.\n\n"
    "Formato: tus respuestas se pintan como Markdown en un terminal estrecho. "
    "Sé conciso, usa bloques de código con su lenguaje y evita tablas anchas.\n"
)

_MACOS_HINTS = (
    "\nEstás en macOS y tienes shell, así que puedes actuar sobre el sistema.\n"
    "Recetas que funcionan:\n"
    "- Crear una carpeta: mkdir -p ruta/de/la/carpeta\n"
    "- Buscar archivos por nombre: mdfind -name 'parte-del-nombre'\n"
    "  (rápido, usa el índice de Spotlight; para un proyecto concreto usa `rg --files`)\n"
    "- Buscar texto dentro de archivos: rg -n 'texto' ruta/\n"
    "- Abrir una app: open -a \"Safari\"\n"
    "- Abrir una web: open 'https://ejemplo.com'\n"
    "- Buscar en Google: open 'https://www.google.com/search?q=lo+que+sea'\n"
    "- Música (Spotify o Music, el que esté abierto):\n"
    "    osascript -e 'tell application \"Spotify\" to playpause'\n"
    "    osascript -e 'tell application \"Spotify\" to next track'\n"
    "    osascript -e 'tell application \"Spotify\" to name of current track'\n"
    "- Volumen: osascript -e 'set volume output volume 50'\n\n"
    "Cuando el usuario te pida una ruta, dásela completa y absoluta.\n"
    "Antes de borrar o sobrescribir algo, pregunta.\n"
)


def build_system_context(lang: str, *, macos: bool = True) -> str:
    """Prompt de sistema que Term aporta en cada turno."""
    lang_name = LANGUAGES.get(lang, LANGUAGES["es"])
    parts = [_SYSTEM_BASE]
    if macos:
        parts.append(_MACOS_HINTS)
    parts.append(f"\nIMPORTANTE: responde siempre en {lang_name} ({lang}).")
    return "".join(parts)

File: frontend/src/commands.ts
/**
 * Ejecución de los comandos que empiezan por «/».
 *
 * Hasta ahora la web los listaba en la ayuda pero no los ejecutaba: escribir
 * /clear se lo mandaba a la IA como si fuera una frase.
 *
 * El catálogo se genera desde el de la terminal (scripts/gen_commands.py), así
 * que aquí solo está lo que el navegador sabe hacer de verdad.
 */

import { COMMAND_GROUPS, WEB_COMMANDS } from './commands.generated'
import { themes, type ThemeKey } from './themes'

export type PanelKey = 'chat' | 'settings' | 'apps' | 'tools' | 'help'

export const EFFORT_LEVELS = ['low', 'medium', 'high', 'max'] as const
export type Effort = (typeof EFFORT_LEVELS)[number]

export const MODELS: Record<string, string> = {
  claude: 'Claude',
  'claude-opus': 'Claude Opus',
  'claude-haiku': 'Claude Haiku',
}

/** Lo que un comando puede pedirle a la aplicación. */
export interface CommandContext {
  addTab: (name?: string, model?: string) => void
  closeTab: () => void
  clearTab: () => void
  renameTab: (name: string) => void
  setPanel: (panel: PanelKey) => void
  setTheme: (theme: ThemeKey) => void
  setEffort: (effort: Effort) => void
  setModel: (model: string) => void
  setWorkdir: (path: string) => void
  copyLast: () => Promise<boolean>
  exportChat: () => void
  search: (text: string) => number
  notify: (text: string) => void
  state: {
    theme: ThemeKey
    effort: Effort
    model: string
    workdir: string
    messages: number
  }
}

export interface CommandOutcome {
  /** Si era un comando; false significa «esto va a la IA». */
  handled: boolean
  /** Texto para enseñar en el chat como respuesta del propio Term. */
  reply?: string
}

/** Sugerencias mientras se escribe, para el desplegable del chat. */
export function suggest(text: string): { cmd: string; desc: string }[] {
  const escrito = text.trim().toLowerCase()
  if (!escrito.startsWith('/')) return []
  const cabeza = escrito.split(/\s+/)[0]
  return COMMAND_GROUPS.flatMap((g) => g.commands)
    .filter((c) => c.web && (escrito === '/' || c.cmd.split(' ')[0].startsWith(cabeza)))
    .slice(0, 8)
    .map((c) => ({ cmd: c.cmd, desc: c.desc }))
}

/** Completa un comando a medio escribir hasta su prefijo común. */
export function complete(text: string): string {
  const escrito = text.trim()
  if (!escrito.startsWith('/')) return text
  const candidatos = WEB_COMMANDS.filter((c) => c.startsWith(escrito))
  if (candidatos.length === 0) return text
  if (candidatos.length === 1) return `${candidatos[0]} `
  let comun = candidatos[0]
  for (const c of candidatos.slice(1)) {
    while (!c.startsWith(comun)) comun = comun.slice(0, -1)
  }
  return comun
}

const listaDeComandos = () =>
  COMMAND_GROUPS.map((g) => {
    const propios = g.commands.filter((c) => c.web)
    if (propios.length === 0) return ''
    const filas = propios.map((c) => `  \`${c.cmd}\` — ${c.desc}`).join('\n')
    return `**${g.title}**\n${filas}`
  })
    .filter(Boolean)
    .join('\n\n')

/**
 * Ejecuta un comando. Devuelve `handled: false` si no lo es, para que el texto
 * siga su camino hacia la IA.
 */
export function runCommand(entrada: string, ctx: CommandContext): CommandOutcome {
  const texto = entrada.trim()
  if (!texto.startsWith('/')) return { handled: false }

  const [cabeza, ...resto] = texto.split(/\s+/)
  const cmd = cabeza.toLowerCase()
  const arg = resto.join(' ').trim()

  switch (cmd) {
    case '/':
      return { handled: true, reply: listaDeComandos() }

    // -- conversación
    case '/new':
      ctx.addTab(arg || undefined)
      return { handled: true }
    case '/close':
      ctx.closeTab()
      return { handled: true }
    case '/clear':
      ctx.clearTab()
      return { handled: true }
    case '/name':
      if (!arg) return { handled: true, reply: 'Uso: `/name <texto>`' }
      ctx.renameTab(arg)
      return { handled: true }
    case '/search': {
      if (!arg) return { handled: true, reply: 'Uso: `/search <texto>`' }
      const n = ctx.search(arg)
      return {
        handled: true,
        reply: n
          ? `${n} ${n === 1 ? 'coincidencia' : 'coincidencias'} de «${arg}»`
          : `Sin coincidencias de «${arg}»`,
      }
    }
    case '/export':
      ctx.exportChat()
      return { handled: true, reply: 'Conversación descargada.' }
    case '/copy':
      void ctx.copyLast().then((ok) =>
        ctx.notify(ok ? 'Copiado al portapapeles' : 'No hay respuesta que copiar'),
      )
      return { handled: true }

    // -- configuración
    case '/model':
      if (!arg) {
        return { handled: true, reply: `Modelos: ${Object.keys(MODELS).join(', ')}` }
      }
      if (!(arg in MODELS)) {
        return { handled: true, reply: `Modelo desconocido: \`${arg}\`` }
      }
      ctx.setModel(arg)
      return { handled: true }
    case '/effort':
      if (!EFFORT_LEVELS.includes(arg as Effort)) {
        return { handled: true, reply: `Niveles: ${EFFORT_LEVELS.join(', ')}` }
      }
      ctx.setEffort(arg as Effort)
      return { handled: true }
    case '/theme':
      if (!(arg in themes)) {
        return { handled: true, reply: `Temas: ${Object.keys(themes).join(', ')}` }
      }
      ctx.setTheme(arg as ThemeKey)
      return { handled: true }
    case '/workdir':
      if (!arg) return { handled: true, reply: `Directorio: \`${ctx.state.workdir || '~'}\`` }
      ctx.setWorkdir(arg)
      return { handled: true }
    case '/save':
      return { handled: true, reply: 'Los ajustes se guardan solos en este navegador.' }

    // -- paneles
    case '/help':
    case '/settings':
    case '/apps':
    case '/tools':
      ctx.setPanel(cmd.slice(1) as PanelKey)
      return { handled: true }
    case '/files':
    case '/map':
      ctx.setPanel('apps')
      return { handled: true }

    // -- información
    case '/tab':
      return {
        handled: true,
        reply:
          `Modelo: **${MODELS[ctx.state.model] ?? ctx.state.model}** · ` +
          `Esfuerzo: **${ctx.state.effort}** · Tema: **${themes[ctx.state.theme].name}** · ` +
          `${ctx.state.messages} mensajes`,
      }
    case '/about':
    case '/version':
      return {
        handled: true,
        reply:
          'Term en el navegador. La versión de terminal (`term`) trae además ' +
          'git, procesos en segundo plano, MCP y control del sistema.',
      }

    default: {
      // Existe en la terminal pero aquí no: mejor decirlo que mandarlo a la IA.
      const enTerminal = COMMAND_GROUPS.flatMap((g) => g.commands).find(
        (c) => c.cmd.split(' ')[0] === cmd,
      )
      if (enTerminal) {
        return {
          handled: true,
          reply: `\`${cmd}\` solo funciona en la terminal, con \`term\`.`,
        }
      }
      return { handled: true, reply: `Comando desconocido: \`${cmd}\`. Prueba \`/help\`.` }
    }
  }
}
